File: quant_system/dashboard/components/js/cooling_panel/cooling_panel.js
// cooling_panel.js
// Circular countdown ring with color transitions
// Handles events: cooling_update, cooling_off

window.CoolingPanel = (function () {

    let container = null;
    let ring = null;
    let label = null;
    let timer = null;

    function init(targetElementId) {
        container = document.getElementById(targetElementId);

        if (!container) return;

        container.innerHTML = `
            <div id="cooling-circle">
                <svg id="cool-svg" width="120" height="120">
                    <circle id="cool-bg" cx="60" cy="60" r="50" stroke="#222" stroke-width="8" fill="none" />
                    <circle id="cool-fg" cx="60" cy="60" r="50" stroke="#4CAF50" stroke-width="8"
                            fill="none" stroke-linecap="round"
                            stroke-dasharray="314" stroke-dashoffset="0" />
                </svg>
                <div id="cool-label">READY</div>
            </div>
        `;

        ring = document.getElementById("cool-fg");
        label = document.getElementById("cool-label");
    }

    function update(payload) {
        const { cool_start, cool_end, remaining_sec } = payload;

        const end = new Date(cool_end).getTime();
        const now = Date.now();
        const total = (end - new Date(cool_start).getTime()) / 1000;
        const remaining = Math.max(remaining_sec, 0);

        const pct = remaining / total;
        const dashOffset = 314 - (314 * pct);
        ring.style.strokeDashoffset = dashOffset;

        let color = "#4CAF50"; // green
        if (pct < 0.66) color = "#FFC400"; // yellow
        if (pct < 0.33) color = "#D50000"; // red

        ring.style.stroke = color;
        label.textContent = `${Math.ceil(remaining)}s`;
    }

    function coolingOff() {
        ring.style.strokeDashoffset = 0;
        ring.style.stroke = "#4CAF50";
        label.textContent = "READY";
    }

    function dispatch(event) {
        switch (event.type) {
            case "cooling_update": return update(event.payload);
            case "cooling_off": return coolingOff();
        }
    }

    return { init, dispatch };

})();

File: quant_system/dashboard/components/js/cooling_panel/cooling_panel.py
"""
cooling_panel.py

Python → JS bridge for cooling countdown visual.
Sends:
 - cooling_start timestamp
 - cooling_end timestamp
 - remaining seconds
"""

from typing import Dict, Any

class CoolingPanelBridge:

    def __init__(self):
        self.js_callback = None

    def register_js_callback(self, cb):
        self.js_callback = cb

    def _send(self, event_type: str, payload: Dict[str, Any]):
        if self.js_callback:
            self.js_callback({"type": event_type, "payload": payload})

    def push_cooling_update(self, start_ts: str, end_ts: str, remaining_sec: int):
        self._send("cooling_update", {
            "cool_start": start_ts,
            "cool_end": end_ts,
            "remaining_sec": remaining_sec
        })

    def push_cooling_off(self):
        self._send("cooling_off", {})
